Open Source Thumbservice

mgdaily opened the pull request:
This is in preparation to open source the thumbservice project. 

A few highlights:
* Added poetry and instructions for developing/testing with it. Also updated docker image to 3.10 and a bunch of core dependencies
* Broke out some LCO-specific configuration items into environment variables
* Updated to use the new archive API fields
* Re-structured the project a bit - there's now a thumbservice "package" which contains all of the modules needed. This is to be more in-line with other projects.

Tested this on my minikube with a test bucket, things are working well. Plan is to push this to a new remote in the OCS org, then archive this project. I'll have to update our helm chart to reference the new docker image location.

Helm charts PR: https://github.com/LCOGT/helm-charts/pull/92
eks-production1 PR: https://github.com/LCOGT/eks-production1/pull/51

Commit:
Migrate dependency management to poetry.

Update docker base image, upgrade dependencies as needed. Fix up docker build to use poetry. Remove requirements.txt

Commit:
Fix up project structure

Commit:
Add configurable settings for open source. Update to use new archive keys

Fix up tests.

Commit:
Add github actions for testing and docker image builds

Commit:
Fix up readme to include poetry

Commit:
Update version of numpy to work with python 3.10

Commit:
Revert "Update version of numpy to work with python 3.10"

This reverts commit fe3cbe62ef333f508faf2016df9b4cf4d6db5dd2.

Commit:
Fixes for python3.10

Commit:
Add readme section about config

Adjust some settings defaults. Remove default for LCO science archive. Also remove endpoint_url setting as it was never used - we always didn't set it and it defaulted to None, which is what boto3 defaults to as well.

Commit:
Update to flask 2

Commit:
Fix mock import

Commit:
Update moto to be python3.10 compatible. Fix tests

For some reason specifying the bucket region doesn't always play well with moto, so remove that. It shouldn't matter anyway - these are just tests.

Commit:
Add some more metadata to pyproject toml file

Commit:
Remove slashes from copy directive. Not needed

Commit:
Remove LCO Jenkinsfile

Commit:
Rename thumbservice to ocs-thumbnail-service

mgdaily commented on thumbservice/common.py:
```diff
@@ -12,13 +12,14 @@ class Settings:
         self.AWS_SECRET_ACCESS_KEY = self.set_value('AWS_SECRET_ACCESS_KEY', 'changeme')
-        # Using `None` for `STORAGE_URL` will connect to AWS
-        self.STORAGE_URL = self.set_value('STORAGE_URL', None)
+        self.REQUIRED_FRAME_VALIDATION_KEYS = self.get_tuple_from_environment('REQUIRED_FRAME_VALIDATION_KEYS', 'configuration_type,request_id,filename')
```
I figured I'd just leave these as sensible defaults, but I could also do away with them and set them in our helm chart.

jnation3406 replied:
I agree, I don't see any harm keeping them. It gives users an idea of what those fields are holding too.

mgdaily commented on thumbservice/tests.py:
```diff
@@ -8,86 +8,86 @@
-        'PROPID': 'LCOEPO2018B-002',
-        'REQNUM': 1756835,
-        'FILTER': 'B'
+        'proposal_id': 'LCOEPO2018B-002',
```
updates to test against new archive API fields

mgdaily commented on thumbservice/tests.py:
```diff
@@ -164,7 +164,7 @@ def thumbservice_client():
 def s3_client():
     # This should be passed in to all test functions to mock out calls to aws
     with mock_s3():
-        config = boto3.session.Config(region_name='us-west-2', signature_version='s3v4')
```
removed this region name because a later version of moto (what we use to mock s3) broke when specifying it. figured it didn't matter much since these are just tests. I've tested the application with a live dev bucket anyway.

jnation3406 replied:
yeah that seems fine... you don't need the region name anyway if you specify the AWS_DEFAULT_REGION env variable.

jnation3406 commented on README.md:
```diff
@@ -12,13 +12,36 @@ set.
+
+|Environment Variable| Description | Default |
+|--------------------|-------------|---------|
+| `ARCHIVE_API` | Archive API URL | http://localhost/
```
should we rename this ARCHIVE_API_URL?

jnation3406 commented on README.md:
```diff
@@ -12,13 +12,36 @@ set.
+|--------------------|-------------|---------|
+| `ARCHIVE_API` | Archive API URL | http://localhost/
+| `TMP_DIR` | Temporary directory for image operations | '/tmp/'
+| `AWS_S3_BUCKET` | AWS S3 Bucket to store thumbnails | 'changeme'
```
maybe just AWS_BUCKET to match archive env variable

jnation3406 commented on thumbservice/thumbservice.py:
```diff
@@ -130,7 +125,6 @@ def get_s3_client():
         's3',
         aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
         aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
-        endpoint_url=settings.STORAGE_URL,
```
I think maybe we need to keep the endpoint URL for S3 in there, so that we can point to a minio server instead when we integrate this into the ocs_example project?

jnation3406 commented on thumbservice/thumbservice.py:
```diff
@@ -164,16 +158,16 @@ def key_exists(key):
 
 def rvb_frames(frames):
-    FILTERS = {
+    PRIMARY_OPTICAL_ELEMENTS = {
```
this is really nitpicky, but this could probably stay as FILTERS or FILTERS_BY_COLOR or something, since it is just a mapping between filter color and filter code. You would still compare it to the primary optical element in the frame

jnation3406 reviewed (approved):
Looks good to me, a few suggestions on env variables.

Commit:
Fix up some environment variable names

Commit:
Add configurable storage url back in

This is for when we use minio, or some other mock s3 server instead of s3 itself.

Commit:
Rename filters variable for clarity, since it's actually talking about filters as a primary optical element
